update indexes.md (#1629)
meilisearch/documentation: learn/core_concepts/indexes.md

## Primary key

> Every index has a primary key: a required attribute that must be present in all documents in the index. The primary key's attribute name must be known by the index. Each document in the index must have a unique value associated with the primary key. This serves to identify each document in the index, such that two documents in an index can never be completely identical. If you try adding documents to an index and even a single one is missing the primary key, none of the documents will be stored.

Author: That's a lot of `index`es

You can set a primary key for an index or let it be inferred by Meilisearch. Read more about it in our [dedicated guide](/learn/core_concepts/primary_key.md#setting-the-primary-key).

[Learn more about document primary key](/learn/core_concepts/primary_key.md#primary-key-2)

## Index settings

Index settings can be thought of as a JSON object containing different options for customizing search behavior.

Meilisearch allows you to customize the following index settings:

- [Ranking rules](#ranking-rules)
- [Synonyms](#synonyms)
- [Filterable attributes](#filterable-attributes)
- [Sortable attributes](#sortable-attributes)
- [Stop words](#stop-words)
- [Displayed and searchable attributes](#displayed-and-searchable-attributes)

> - [Typo tolerance](#typo-tolerance)

Author: Oh wow, I completely forgot about it 😅

> These settings are what differentiate between indexes even if they have the same documents. You can make these changes using the [settings route](/reference/api/settings.md) or the [dedicated settings child route](/reference/api/settings.md#all-settings).

Reviewer: I'm trying to shift the emphasis from "settings differentiate/distinguish indexes" to "settings make indexes behave differently." Like, I feel that the differentiation itself is less important than the reason why you would make things different in the first place.

```suggestion
These settings can make indexes behave differently even if they have the same documents. To change them, you can use the [settings route](/reference/api/settings.md) or the [dedicated settings child route](/reference/api/settings.md#all-settings).
```

Since synonyms are defined for a given index, they won't apply to any other index on the same Meilisearch instance. You can create your list of synonyms using the [update settings endpoint](/reference/api/settings.md#update-settings) or the [update synonyms endpoint](/reference/api/synonyms.md#update-synonyms).

[Learn more about synonyms](/learn/configuration/synonyms.md)

### Filterable attributes

Filtering allows you to refine your search based on different categories, in our case, document attributes; these can include a movie genre, a price range, or location. To filter by any document attribute, you need to add it to `filterableAttributes` using the [update settings endpoint](/reference/api/settings.md#update-settings) or the [update filterable attributes endpoint](/reference/api/filterable_attributes.md#update-filterable-attributes).

Adding an attribute to `filterableAttributes` will not impact search. You need it to use the [`filter` search parameter](/reference/api/search.md#filter) to refine your results.

[Learn more about filtering](/learn/advanced/filtering_and_faceted_search.md)

### Sortable attributes

> By default, Meilisearch orders results according to their relevancy. You can alter this sorting behavior to show certain results first by adding attributes to the `sortableAttributes` array using the [update settings endpoint](/reference/api/settings.md#update-settings) or the [update sortable attributes endpoint](/reference/api/sortable_attributes.md#update-sortable-attributes). You can then use the [`sort` search parameter](/reference/api/search.md#sort) to sort your results in ascending or descending oder.

Reviewer 2: ```suggestion
By default, Meilisearch orders results according to their relevancy. You can alter this sorting behavior to show certain results first.

Add the attributes you'd like to sort by to `sortableAttributes` using the [update settings endpoint](/reference/api/settings.md#update-settings) or the [update sortable attributes endpoint](/reference/api/sortable_attributes.md#update-sortable-attributes). You can then use the [`sort` search parameter](/reference/api/search.md#sort) to sort your results in ascending or descending order.
```

[Learn more about sorting](/learn/advanced/sorting.md)

### Stop words

> Your dataset may contain words you want to ignore during search. These may include words that don't add semantic value or context or are too frequent (e.g., `the` or `of` in English). You can add such words to the [stop words list](/reference/api/stop_words.md) and Meilisearch will ignore them during search. In addition to improving relevancy, designating common words as stop words also greatly improves performance.

Reviewer 2: ```suggestion
Your dataset may contain words you want to ignore during search because, for example, they don't add semantic value or occur too frequently (e.g., `the` or `of` in English). You can add these words to the [stop words list](/reference/api/stop_words.md) and Meilisearch will ignore them during search.
```

> You can create your list of stop words for your index using the [update settings endpoint](/reference/api/settings.md#update-settings) or the [update stop words endpoint](/reference/api/stop_words.md#update-stop-words).

Reviewer 2: ```suggestion
Change your index's stop words list using the [update settings endpoint](/reference/api/settings.md#update-settings) or the [update stop words endpoint](/reference/api/stop_words.md#update-stop-words). In addition to improving relevancy, designating common words as stop words greatly improves performance.
```

[Learn more about stop words](/reference/api/stop_words.md)

### Displayed and searchable attributes

By default, every document field is searchable and displayed in response to search queries. However, you can choose to set some fields as non-searchable, non-displayed, or both.

> You can update these field attributes using the [update settings endpoint](/reference/api/settings.md#update-settings), or the respective update endpoints for [displayed attributes](/reference/api/displayed_attributes.md#update-displayed-attributes), and [searchable attributes](/reference/api/searchable_attributes.md#update-searchable-attributes).

Reviewer 2: ```suggestion
You can update these field attributes using the [update settings endpoint](/reference/api/settings.md#update-settings), or the respective endpoints for [displayed attributes](/reference/api/displayed_attributes.md#update-displayed-attributes) and [searchable attributes](/reference/api/searchable_attributes.md#update-searchable-attributes).
```

[Learn more about displayed and searchable attributes](/learn/configuration/displayed_searchable_attributes.md)

### Typo tolerance

> Typo tolerance is a built-in feature that helps you find relevant results even when your search queries contain spelling mistakes or typos, e.g., typing `chickne` instead of `chicken`. The setting allows you to configure the following for an index:

Reviewer 2: ```suggestion
Typo tolerance is a built-in feature that helps you find relevant results even when your search queries contain spelling mistakes or typos, e.g., typing `chickne` instead of `chicken`. This setting allows you to do the following for your index:
```

- Enable or disable typo tolerance
- Configure the minimum word size for typos
- Disable typos on specific words
- Disable typos on specific document attributes

> You can update the typo tolerance settings for an index using the [update settings endpoint](/reference/api/settings.md#update-settings) or the [update typo tolerance endpoint](/reference/api/typo_tolerance.md#update-typo-tolerance).

Reviewer 2: ```suggestion
You can update the typo tolerance settings using the [update settings endpoint](/reference/api/settings.md#update-settings) or the [update typo tolerance endpoint](/reference/api/typo_tolerance.md#update-typo-tolerance).
```

- [Ranking rules](#ranking-rules)
- [Distinct attribute](#distinct-attribute)
- [Synonyms](#synonyms)
- [Filterable attributes](#filterable-attributes)
- [Sortable attributes](#sortable-attributes)
- [Stop words](#stop-words)
- [Displayed and searchable attributes](#displayed-and-searchable-attributes)
- [Typo tolerance](#typo-tolerance)

To change index settings, use the [update settings endpoint](/reference/api/settings.md#update-settings) or any of the [child routes](/reference/api/settings.md#all-settings).

### Ranking rules

All indexes are created with the same built-in ranking rules executed in default order:

```
[
  "words",
  "typo",
  "proximity",
  "attribute",
  "sort",
  "exactness"
]
```

The order of these rules matters: the first rule has the most impact, and the last rule has the least. You can alter this order or define custom ranking rules to return certain results first. This can be done using the [update settings endpoint](/reference/api/settings.md#update-settings) or the [update ranking rules endpoint](/reference/api/ranking_rules.md#update-ranking-rules).

[Learn more about ranking rules](/learn/core_concepts/relevancy.md)

### Distinct attribute

> If your dataset contains multiple similar documents, you may want to return only one on search. Suppose you have numerous black jackets in different sizes in your `costumes` index; setting `costume_id` as the distinct attribute will mean Meiliserch will not return more than one black jacket with the same `costume_id`.

Reviewer 2: Changed the example from `costume_id` to `costume_name`. Since `costume_id` contains `id` it can read as a primary key, in which case there wouldn't be two documents w/ the same `costume_id` and making it distinct attribute would be meaningless.

Not that most readers would understand that distinction at this point. It's more for me and my peace of mind 😂 

```suggestion
If your dataset contains multiple similar documents, you may want to return only one on search. Suppose you have numerous black jackets in different sizes in your `costumes` index; setting `costume_name` as the distinct attribute will mean Meiliserch will not return more than one black jacket with the same `costume_name`.
```

> You can only set one field as the distinct attribute per index using the [update settings endpoint](/reference/api/settings.md#update-settings) or the [update distinct attribute endpoint](/reference/api/distinct_attribute.md#update-distinct-attribute)

Reviewer 2: ```suggestion
Designate the distinct attribute using the [update settings endpoint](/reference/api/settings.md#update-settings) or the [update distinct attribute endpoint](/reference/api/distinct_attribute.md#update-distinct-attribute). **You can only set one field as the distinct attribute per index.** 
```

[Learn more about distinct attribute](/learn/configuration/distinct.md)

### Synonyms

Your dataset may contain words with similar meanings. For these, you can define a list of synonyms: words that will be treated as the same or similar for search purposes.

Since synonyms are defined for a given index, they won't apply to any other index on the same Meilisearch instance. You can create your list of synonyms using the [update settings endpoint](/reference/api/settings.md#update-settings) or the [update synonyms endpoint](/reference/api/synonyms.md#update-synonyms).

[Learn more about synonyms](/learn/configuration/synonyms.md)

### Filterable attributes

> Filtering allows you to refine your search based on different categories. For example, you could search for all movies of a certain `genre`, e.g. `Science Fiction`, with a `rating` above `8`.

Author: ```suggestion
Filtering allows you to refine your search based on different categories. For example, you could search for all movies of a certain `genre`, e.g., `Science Fiction`, with a `rating` above `8`.
```
